// ng-shopping-cart/__ivy_ngcc__/esm2015/ng-shopping-cart.js
import { EventEmitter, Component, Input, Output, Inject, LOCALE_ID, ComponentFactoryResolver, Directive, NgModuleRef, ViewContainerRef, InjectionToken, Injectable, Pipe, NgModule } from '@angular/core';
import { CurrencyPipe, getLocaleCurrencyName, CommonModule } from '@angular/common';
import { HttpClient, HttpParams, HttpRequest, HttpClientModule } from '@angular/common/http';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Converts a string into an object with all currency display options as properties.
 * @param {?} format
 * @return {?}
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@angular/common/http';

function AddToCartComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelementStart(1, "add-to-cart-editor", 5);
    ɵngcc0.ɵɵlistener("valueChange", function AddToCartComponent_div_1_Template_add_to_cart_editor_valueChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r5); const ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.editorQuantity = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r0.position);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("type", ctx_r0.type)("dropdown", ctx_r0.dropdown)("value", ctx_r0.editorQuantity);
} }
function AddToCartComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "button", 6);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r1.buttonClass)("disabled", !ctx_r1.item);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r1.buttonText, " ");
} }
function AddToCartComponent_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0, 0, ["*ngIf", "custom"]);
} }
function AddToCartComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelementStart(1, "add-to-cart-editor", 5);
    ɵngcc0.ɵɵlistener("valueChange", function AddToCartComponent_div_5_Template_add_to_cart_editor_valueChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.editorQuantity = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r3.position);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("type", ctx_r3.type)("dropdown", ctx_r3.dropdown)("value", ctx_r3.editorQuantity);
} }
const _c0 = ["*"];
function CartCheckoutComponent_ng_container_0_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 3);
    ɵngcc0.ɵɵlistener("click", function CartCheckoutComponent_ng_container_0_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r5); const ctx_r4 = ɵngcc0.ɵɵnextContext(2); return ctx_r4.doCheckout(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r2.buttonClass)("disabled", ctx_r2.empty);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r2.buttonText, " ");
} }
function CartCheckoutComponent_ng_container_0_span_2_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "span", 4);
    ɵngcc0.ɵɵlistener("click", function CartCheckoutComponent_ng_container_0_span_2_Template_span_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(2); return ctx_r6.doCheckout(); });
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementEnd();
} }
function CartCheckoutComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, CartCheckoutComponent_ng_container_0_button_1_Template, 2, 3, "button", 1);
    ɵngcc0.ɵɵtemplate(2, CartCheckoutComponent_ng_container_0_span_2_Template, 2, 0, "span", 2);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.custom);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.custom);
} }
function CartCheckoutComponent_ng_container_1_input_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "input", 18);
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("value", ctx_r8.paypalSettings.itemName);
} }
function CartCheckoutComponent_ng_container_1_input_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "input", 19);
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("value", ctx_r9.paypalSettings.itemNumber);
} }
function CartCheckoutComponent_ng_container_1_input_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "input", 20);
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("value", ctx_r10.paypalSettings.serviceName + "_BuyNow_WPS_" + ctx_r10.paypalSettings.country);
} }
function CartCheckoutComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "form", 5);
    ɵngcc0.ɵɵelement(2, "input", 6);
    ɵngcc0.ɵɵelement(3, "input", 7);
    ɵngcc0.ɵɵelement(4, "input", 8);
    ɵngcc0.ɵɵtemplate(5, CartCheckoutComponent_ng_container_1_input_5_Template, 1, 1, "input", 9);
    ɵngcc0.ɵɵtemplate(6, CartCheckoutComponent_ng_container_1_input_6_Template, 1, 1, "input", 10);
    ɵngcc0.ɵɵelement(7, "input", 11);
    ɵngcc0.ɵɵelement(8, "input", 12);
    ɵngcc0.ɵɵelement(9, "input", 13);
    ɵngcc0.ɵɵelement(10, "input", 14);
    ɵngcc0.ɵɵtemplate(11, CartCheckoutComponent_ng_container_1_input_11_Template, 1, 1, "input", 15);
    ɵngcc0.ɵɵelement(12, "input", 16);
    ɵngcc0.ɵɵelement(13, "img", 17);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("value", ctx_r1.paypalSettings.business);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("value", ctx_r1.paypalLocale);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.paypalSettings.itemName);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.paypalSettings.itemNumber);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("value", ctx_r1.cost);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("value", ctx_r1.currency);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("value", ctx_r1.taxRate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("value", ctx_r1.shipping);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.paypalSettings.serviceName && ctx_r1.paypalSettings.country);
} }
function CartSummaryComponent__svg_svg_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(0, "svg", 7);
    ɵngcc0.ɵɵelement(1, "path", 8);
    ɵngcc0.ɵɵelement(2, "circle", 9);
    ɵngcc0.ɵɵelement(3, "circle", 10);
    ɵngcc0.ɵɵelementEnd();
} }
function CartSummaryComponent_img_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 11);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("src", ctx_r1.icon, ɵngcc0.ɵɵsanitizeUrl);
} }
const _c1 = function (a0, a1, a2, a3, a4) { return [a0, a1, a2, a3, a4]; };
function AddToCartEditorComponent_ng_container_0_option_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "option", 4);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r4 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("selected", item_r4.value === ctx_r3.value)("value", item_r4.value);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(item_r4.label);
} }
function AddToCartEditorComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "select", 1, 2);
    ɵngcc0.ɵɵlistener("change", function AddToCartEditorComponent_ng_container_0_Template_select_change_1_listener() { ɵngcc0.ɵɵrestoreView(_r6); const _r2 = ɵngcc0.ɵɵreference(2); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.changeValue(_r2.value); });
    ɵngcc0.ɵɵtemplate(3, AddToCartEditorComponent_ng_container_0_option_3_Template, 2, 3, "option", 3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.dropdown);
} }
function AddToCartEditorComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "input", 5, 6);
    ɵngcc0.ɵɵlistener("change", function AddToCartEditorComponent_ng_container_1_Template_input_change_1_listener() { ɵngcc0.ɵɵrestoreView(_r9); const _r7 = ɵngcc0.ɵɵreference(2); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.changeValue(_r7.value); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("type", ctx_r1.type)("value", ctx_r1.value);
} }
function CartViewComponent_ng_container_0_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r2.emptyText, " ");
} }
function CartViewComponent_ng_container_0_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0, 0, ["*ngIf", "customEmptyContent"]);
} }
function CartViewComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, CartViewComponent_ng_container_0_div_1_Template, 2, 1, "div", 1);
    ɵngcc0.ɵɵtemplate(2, CartViewComponent_ng_container_0_2_Template, 1, 0, undefined, 0);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.customEmptyContent);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.customEmptyContent);
} }
const _c2 = function (a0) { return { "background-image": a0 }; };
function CartViewComponent_ng_container_1_ng_container_1_div_1_div_12_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 20);
    ɵngcc0.ɵɵelement(1, "div", 21);
    ɵngcc0.ɵɵelementStart(2, "div", 22);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 23);
    ɵngcc0.ɵɵelementStart(5, "button", 24);
    ɵngcc0.ɵɵlistener("click", function CartViewComponent_ng_container_1_ng_container_1_div_1_div_12_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r13); const cartItem_r11 = ctx.$implicit; const ctx_r12 = ɵngcc0.ɵɵnextContext(4); return ctx_r12.increase(cartItem_r11); });
    ɵngcc0.ɵɵtext(6, " + ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "span", 25);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "button", 26);
    ɵngcc0.ɵɵlistener("click", function CartViewComponent_ng_container_1_ng_container_1_div_1_div_12_Template_button_click_9_listener() { ɵngcc0.ɵɵrestoreView(_r13); const cartItem_r11 = ctx.$implicit; const ctx_r14 = ɵngcc0.ɵɵnextContext(4); return ctx_r14.decrease(cartItem_r11); });
    ɵngcc0.ɵɵtext(10, " - ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 27);
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "currency");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "div", 28);
    ɵngcc0.ɵɵtext(15);
    ɵngcc0.ɵɵpipe(16, "currency");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const cartItem_r11 = ctx.$implicit;
    const ctx_r7 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(17, _c2, "url(" + cartItem_r11.getImage() + ")"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", cartItem_r11.getName(), " ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(cartItem_r11.getQuantity());
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBindV(13, 5, ɵngcc0.ɵɵpureFunction5(19, _c1, cartItem_r11.getPrice(), ctx_r7.format.currencyCode, ctx_r7.format.display, ctx_r7.format.digitsInfo, ctx_r7.format.locale)), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBindV(16, 11, ɵngcc0.ɵɵpureFunction5(25, _c1, cartItem_r11.total(), ctx_r7.format.currencyCode, ctx_r7.format.display, ctx_r7.format.digitsInfo, ctx_r7.format.locale)), " ");
} }
function CartViewComponent_ng_container_1_ng_container_1_div_1_div_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 16);
} }
function CartViewComponent_ng_container_1_ng_container_1_div_1_div_24_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 16);
} }
function CartViewComponent_ng_container_1_ng_container_1_div_1_div_32_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 16);
} }
const _c3 = function (a0) { return { "no-images": a0 }; };
function CartViewComponent_ng_container_1_ng_container_1_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelementStart(1, "div", 5);
    ɵngcc0.ɵɵelementStart(2, "div", 6);
    ɵngcc0.ɵɵelement(3, "div", 7);
    ɵngcc0.ɵɵelementStart(4, "div", 8);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 9);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "div", 10);
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "div", 11);
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(12, CartViewComponent_ng_container_1_ng_container_1_div_1_div_12_Template, 17, 31, "div", 12);
    ɵngcc0.ɵɵelementStart(13, "div", 13);
    ɵngcc0.ɵɵelementStart(14, "div", 14);
    ɵngcc0.ɵɵtemplate(15, CartViewComponent_ng_container_1_ng_container_1_div_1_div_15_Template, 1, 0, "div", 15);
    ɵngcc0.ɵɵelement(16, "div", 16);
    ɵngcc0.ɵɵelement(17, "div", 16);
    ɵngcc0.ɵɵelement(18, "div", 16);
    ɵngcc0.ɵɵelementStart(19, "div", 17);
    ɵngcc0.ɵɵtext(20);
    ɵngcc0.ɵɵpipe(21, "percent");
    ɵngcc0.ɵɵpipe(22, "currency");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(23, "div", 14);
    ɵngcc0.ɵɵtemplate(24, CartViewComponent_ng_container_1_ng_container_1_div_1_div_24_Template, 1, 0, "div", 15);
    ɵngcc0.ɵɵelement(25, "div", 16);
    ɵngcc0.ɵɵelement(26, "div", 16);
    ɵngcc0.ɵɵelement(27, "div", 16);
    ɵngcc0.ɵɵelementStart(28, "div", 18);
    ɵngcc0.ɵɵtext(29);
    ɵngcc0.ɵɵpipe(30, "currency");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(31, "div", 14);
    ɵngcc0.ɵɵtemplate(32, CartViewComponent_ng_container_1_ng_container_1_div_1_div_32_Template, 1, 0, "div", 15);
    ɵngcc0.ɵɵelement(33, "div", 16);
    ɵngcc0.ɵɵelement(34, "div", 16);
    ɵngcc0.ɵɵelement(35, "div", 16);
    ɵngcc0.ɵɵelementStart(36, "div", 19);
    ɵngcc0.ɵɵtext(37);
    ɵngcc0.ɵɵpipe(38, "currency");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r6.display + "-display");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(39, _c3, !ctx_r6.images));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r6.nameHeaderText, " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r6.quantityHeaderText, " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r6.priceHeaderText, " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r6.totalHeaderText, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r6.items);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.images);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate3(" ", ctx_r6.taxFooterText, ": (", ɵngcc0.ɵɵpipeBind3(21, 17, ctx_r6.taxRate, ctx_r6.format.digitsInfo, ctx_r6.format.locale), ") ", ɵngcc0.ɵɵpipeBindV(22, 21, ɵngcc0.ɵɵpureFunction5(41, _c1, ctx_r6.tax, ctx_r6.format.currencyCode, ctx_r6.format.display, ctx_r6.format.digitsInfo, ctx_r6.format.locale)), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.images);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate2(" ", ctx_r6.shippingFooterText, ": ", ɵngcc0.ɵɵpipeBindV(30, 27, ɵngcc0.ɵɵpureFunction5(47, _c1, ctx_r6.shipping, ctx_r6.format.currencyCode, ctx_r6.format.display, ctx_r6.format.digitsInfo, ctx_r6.format.locale)), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.images);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate2(" ", ctx_r6.totalFooterText, ": ", ɵngcc0.ɵɵpipeBindV(38, 33, ɵngcc0.ɵɵpureFunction5(53, _c1, ctx_r6.cost, ctx_r6.format.currencyCode, ctx_r6.format.display, ctx_r6.format.digitsInfo, ctx_r6.format.locale)), " ");
} }
function CartViewComponent_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, CartViewComponent_ng_container_1_ng_container_1_div_1_Template, 39, 59, "div", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r4.empty);
} }
function CartViewComponent_ng_container_1_ng_container_2_th_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "th", 7);
} }
function CartViewComponent_ng_container_1_ng_container_2_tr_14_td_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "td", 21);
} if (rf & 2) {
    const cartItem_r17 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(1, _c2, "url(" + cartItem_r17.getImage() + ")"));
} }
function CartViewComponent_ng_container_1_ng_container_2_tr_14_Template(rf, ctx) { if (rf & 1) {
    const _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tr");
    ɵngcc0.ɵɵtemplate(1, CartViewComponent_ng_container_1_ng_container_2_tr_14_td_1_Template, 1, 3, "td", 32);
    ɵngcc0.ɵɵelementStart(2, "td", 22);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "td", 23);
    ɵngcc0.ɵɵelementStart(5, "button", 24);
    ɵngcc0.ɵɵlistener("click", function CartViewComponent_ng_container_1_ng_container_2_tr_14_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r21); const cartItem_r17 = ctx.$implicit; const ctx_r20 = ɵngcc0.ɵɵnextContext(3); return ctx_r20.increase(cartItem_r17); });
    ɵngcc0.ɵɵtext(6, " + ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "span", 25);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "button", 26);
    ɵngcc0.ɵɵlistener("click", function CartViewComponent_ng_container_1_ng_container_2_tr_14_Template_button_click_9_listener() { ɵngcc0.ɵɵrestoreView(_r21); const cartItem_r17 = ctx.$implicit; const ctx_r22 = ɵngcc0.ɵɵnextContext(3); return ctx_r22.decrease(cartItem_r17); });
    ɵngcc0.ɵɵtext(10, " - ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "td", 27);
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "currency");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "td", 28);
    ɵngcc0.ɵɵtext(15);
    ɵngcc0.ɵɵpipe(16, "currency");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const cartItem_r17 = ctx.$implicit;
    const ctx_r16 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r16.images);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(cartItem_r17.getName());
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(cartItem_r17.getQuantity());
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBindV(13, 5, ɵngcc0.ɵɵpureFunction5(17, _c1, cartItem_r17.getPrice(), ctx_r16.format.currencyCode, ctx_r16.format.display, ctx_r16.format.digitsInfo, ctx_r16.format.locale)), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBindV(16, 11, ɵngcc0.ɵɵpureFunction5(23, _c1, cartItem_r17.total(), ctx_r16.format.currencyCode, ctx_r16.format.display, ctx_r16.format.digitsInfo, ctx_r16.format.locale)), " ");
} }
function CartViewComponent_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "table", 29);
    ɵngcc0.ɵɵelementStart(2, "thead", 6);
    ɵngcc0.ɵɵelementStart(3, "tr");
    ɵngcc0.ɵɵtemplate(4, CartViewComponent_ng_container_1_ng_container_2_th_4_Template, 1, 0, "th", 30);
    ɵngcc0.ɵɵelementStart(5, "th", 8);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "th", 9);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "th", 10);
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "th", 11);
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(13, "tbody");
    ɵngcc0.ɵɵtemplate(14, CartViewComponent_ng_container_1_ng_container_2_tr_14_Template, 17, 29, "tr", 31);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(15, "tfoot", 13);
    ɵngcc0.ɵɵelementStart(16, "tr", 14);
    ɵngcc0.ɵɵelement(17, "td", 16);
    ɵngcc0.ɵɵelementStart(18, "td", 17);
    ɵngcc0.ɵɵtext(19);
    ɵngcc0.ɵɵpipe(20, "percent");
    ɵngcc0.ɵɵpipe(21, "currency");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(22, "tr", 14);
    ɵngcc0.ɵɵelement(23, "td", 16);
    ɵngcc0.ɵɵelementStart(24, "td", 18);
    ɵngcc0.ɵɵtext(25);
    ɵngcc0.ɵɵpipe(26, "currency");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(27, "tr", 14);
    ɵngcc0.ɵɵelement(28, "td", 16);
    ɵngcc0.ɵɵelementStart(29, "td", 19);
    ɵngcc0.ɵɵtext(30);
    ɵngcc0.ɵɵpipe(31, "currency");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.images);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r5.nameHeaderText);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r5.quantityHeaderText);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r5.priceHeaderText);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r5.totalHeaderText);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r5.items);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵattribute("colspan", ctx_r5.images ? 4 : 3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate3(" ", ctx_r5.taxFooterText, ": (", ɵngcc0.ɵɵpipeBind3(20, 16, ctx_r5.taxRate, ctx_r5.format.digitsInfo, ctx_r5.format.locale), ") ", ɵngcc0.ɵɵpipeBindV(21, 20, ɵngcc0.ɵɵpureFunction5(38, _c1, ctx_r5.tax, ctx_r5.format.currencyCode, ctx_r5.format.display, ctx_r5.format.digitsInfo, ctx_r5.format.locale)), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵattribute("colspan", ctx_r5.images ? 4 : 3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate2(" ", ctx_r5.shippingFooterText, ": ", ɵngcc0.ɵɵpipeBindV(26, 26, ɵngcc0.ɵɵpureFunction5(44, _c1, ctx_r5.shipping, ctx_r5.format.currencyCode, ctx_r5.format.display, ctx_r5.format.digitsInfo, ctx_r5.format.locale)), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵattribute("colspan", ctx_r5.images ? 4 : 3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate2(" ", ctx_r5.totalFooterText, ": ", ɵngcc0.ɵɵpipeBindV(31, 32, ɵngcc0.ɵɵpureFunction5(50, _c1, ctx_r5.cost, ctx_r5.format.currencyCode, ctx_r5.format.display, ctx_r5.format.digitsInfo, ctx_r5.format.locale)), " ");
} }
function CartViewComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, CartViewComponent_ng_container_1_ng_container_1_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(2, CartViewComponent_ng_container_1_ng_container_2_Template, 32, 56, "ng-container", 0);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.display !== "table");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.display === "table");
} }
function CartShowcaseComponent_div_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
const _c4 = function (a0, a1, a2, a3, a4, a5) { return [a0, a1, a2, a3, a4, a5]; };
function CartShowcaseComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "div", 3);
    ɵngcc0.ɵɵtemplate(2, CartShowcaseComponent_div_1_ng_container_2_Template, 1, 0, "ng-container", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const carItem_r1 = ctx.$implicit;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction6(6, _c4, ctx_r0.xsClass, ctx_r0.sClass, ctx_r0.mClass, ctx_r0.lClass, ctx_r0.xlClass, ctx_r0.ratioClass));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("cartShowcaseOutlet", ctx_r0.itemComponent)("cartShowcaseOutletItem", carItem_r1)("cartShowcaseOutletFormat", ctx_r0.format)("cartShowcaseOutletInjector", ctx_r0.injector)("cartShowcaseOutletNgModuleFactory", ctx_r0.moduleFactory);
} }
function parseLocaleFormat(format) {
    if (!format) {
        throw new Error('Invalid format for currency. Expected a non empty string');
    }
    const /** @type {?} */ res = { currencyCode: undefined, display: 'symbol', digitsInfo: undefined, locale: undefined };
    const /** @type {?} */ props = ['currencyCode', 'display', 'digitsInfo', 'locale'];
    const /** @type {?} */ parts = format.split(':');
    const /** @type {?} */ partsLength = parts.length;
    if (partsLength > 4) {
        throw new Error(`Invalid format for currency. Expected a value in the form ${props.join(':')} and got ${format}`);
    }
    parts.forEach((p, idx) => {
        let /** @type {?} */ value = p;
        if (value === 'auto') {
            value = idx === 1 ? 'symbol' : undefined;
        }
        res[props[idx]] = value;
    });
    return res;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// unsupported: template constraints.
/**
 * The base class for storing items in your cart
 *
 * \@service
 * \@note {warning} Do not modify the items `id` after they are added to the cart. Doing so could result in duplicates which can cause
 * undefined behaviour
 * \@order 1
 * @abstract
 * @template T
 */
class CartService {
    constructor() {
        this.localeFormat = parseLocaleFormat('auto');
        this.format = 'auto';
        /**
         * Emits an event every time items, tax, shipping cost or currency formats are changed in the cart.
         */
        this.onChange = new EventEmitter();
        /**
         * Emits an event every time an item is added to the cart.
         */
        this.onItemAdded = new EventEmitter();
        /**
         * Emits an event every time an item is removed from the cart.
         *
         * > This event only fires when a single item is removed by id. If you want to be notified of any removal (eg: clearing the cart) listen
         * to the `onChange` or the `onItemsChanged` event instead.
         */
        this.onItemRemoved = new EventEmitter();
        /**
         * Emits an event every time an item is added or removed from the cart.
         */
        this.onItemsChanged = new EventEmitter();
        /**
         * Emits an event every time taxes for the cart are changed.
         */
        this.onTaxChange = new EventEmitter();
        /**
         * Emits an event every time shipping costs for the cart are changed.
         */
        this.onShippingChange = new EventEmitter();
    }
    /**
     * Returns the tax computation of the shopping cart.
     * @return {?}
     */
    getTax() {
        return this.cost() * (this.getTaxRate() / 100);
    }
    /**
     * Returns the total cost of the shopping cart including shipping and taxes.
     * @return {?}
     */
    totalCost() {
        return this.cost() + this.getTax() + this.getShipping();
    }
    /**
     * Changes the currency symbol and number format for all components associated to this service instance.
     *
     * Check the Angular `CurrencyPipe` and the Internationalization guide for more info.
     * @param {?} format
     * @return {?}
     */
    setLocaleFormat(format) {
        this.localeFormat = parseLocaleFormat(format);
        this.format = format;
        this.onChange.emit({ change: 'format', value: this.format });
    }
    /**
     * Returns the currency format as set with `setCurrencyFormat` or `'auto'` if no value is set.
     *
     * Passing true as parameter will return an object instead of a string.
     * @param {?=} object
     * @return {?}
     */
    getLocaleFormat(object = false) {
        return object ? this.localeFormat : this.format;
    }
    /**
     * Returns an object with all the cart information in it, useful for serialization of the cart.
     * @return {?}
     */
    toObject() {
        return {
            taxRate: this.getTaxRate(),
            shipping: this.getShipping(),
            items: this.getItems()
        };
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * The base class for every unit of information stored in the cart service
 * \@order 1
 * @abstract
 */
class CartItem {
    /**
     * Return the total cost of the item, that is the price multiplied by the quantity
     * @return {?}
     */
    total() {
        return this.getPrice() * this.getQuantity();
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * A component to add items to the cart by pressing a button. Has different built-in editors to select quantity.
 *
 * \@order 1
 * \@howToUse "With a custom button or projected content"
 * ```html
 * <add-to-cart [item]="item" [custom]="true">
 *    <button type="button" class="my-custom-class">Add item</button>
 * </add-to-cart>
 * ```
 *
 * \@howToUse "With different text and classes"
 * ```html
 * <add-to-cart [item]="item" [buttonText]="'Add item'" [buttonClass]="'my-custom-class'">
 * </add-to-cart>
 * ```
 *
 * \@howToUse "With a html number input positioned on top"
 * ```html
 * <add-to-cart [item]="item" [type]="'number'" [position]="'top'">
 * </add-to-cart>
 * ```
 *
 * \@howToUse "With a select for selecting quantity"
 * ```html
 * <add-to-cart [item]="item" [type]="'dropdown'" [dropdown]="[{ label: 'One item', value: 1 }, { label: 'Two items', value: 2 }]">
 * </add-to-cart>
 * ```
 *
 * \@howToUse "With the default button and different quantity"
 * ```html
 * <add-to-cart [item]="item" [quantity]="5">
 * </add-to-cart>
 * ```
 *
 * \@note {warning} This component captures click events that bubble from its projected content if you are using `[custom]=true` therefore if
 * you have html content other than buttons inside you must stop the event propagation unless the click originated in the button that
 * add the items to the cart.
 */
class AddToCartComponent {
    /**
     * @param {?} cartService
     */
    constructor(cartService) {
        this.cartService = cartService;
        this._editorQuantity = 1;
        this.hasEditor = false;
        this.horizontalEditor = true;
        this.editorPrecedence = 'before';
        /**
         * If `false` displays a default button provided by the component, otherwise projects the contents of the component to be used as a
         * button.
         */
        this.custom = false;
        /**
         * Changes the default text of the component's button.
         */
        this.buttonText = 'Add to cart';
        /**
         * Changes the default CSS class of the component's button.
         */
        this.buttonClass = 'add-to-cart-button';
        /**
         * Renders a button or a button with an editor to select the quantity of the item that will be added in the cart. When it has a value
         * other than `'button'` an editor is displayed depending on the selected `[type]`; it can be a `select`, or a text or a number `input`.
         *
         * > Do not confuse this input with the html attribute `type`. The default button is always generated with this attribute set to
         * `button` to prevent accidental form submissions.
         */
        this.type = 'button';
        /**
         * Sets the position where the editor will be placed. If the `[type]` is set to `'button'` no editor is displayed and this setting has
         * no effect.
         */
        this.position = 'left';
        /**
         * If `[type]` is set to `'dropdown'` it can be used to set the options of the rendered `select` editor. Is an array of objects with
         * label and a value properties used to populate the select's `option` elements.
         */
        this.dropdown = [
            { label: '1 item', value: 1 },
            { label: '2 items', value: 2 },
            { label: '5 items', value: 5 }
        ];
        /**
         * This event is fired when the component uses an editor and its value is changed by the user.
         */
        this.change = new EventEmitter();
        /**
         * This event is fired when the item is added to the cart.
         */
        this.added = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get editorQuantity() {
        return this._editorQuantity;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set editorQuantity(value) {
        this._editorQuantity = value;
        this.change.emit(value);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.computeClass();
    }
    /**
     * @return {?}
     */
    itemQuantity() {
        if (this.type === 'button') {
            if (this.quantity) {
                return this.quantity;
            }
            return this.item.getQuantity();
        }
        else {
            return this._editorQuantity;
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes['type']) {
            this.hasEditor = changes['type'].currentValue !== 'button';
            if (changes['type'].currentValue === 'dropdown' && this.dropdown.length) {
                const /** @type {?} */ quantity = this.itemQuantity();
                const /** @type {?} */ match = this.dropdown.find(i => i.value === quantity);
                if (!match) {
                    this._editorQuantity = this.dropdown[0].value;
                }
            }
        }
        if (changes['position']) {
            const /** @type {?} */ pos = changes['position'].currentValue;
            this.horizontalEditor = pos === 'left' || pos === 'right';
            this.editorPrecedence = pos === 'left' || pos === 'top' ? 'before' : 'after';
        }
        this.computeClass();
    }
    /**
     * @param {?} evt
     * @return {?}
     */
    addToCart(evt) {
        evt.stopPropagation();
        if (this.item) {
            const /** @type {?} */ quantity = this.itemQuantity();
            this.item.setQuantity(quantity);
            this.cartService.addItem(this.item);
            this.added.emit(this.item);
        }
    }
    /**
     * @return {?}
     */
    computeClass() {
        this.containerClass = [
            'add-to-cart-' + this.type,
            this.horizontalEditor ?
                'editor-position-horizontal' :
                'editor-position-vertical'
        ];
    }
}
AddToCartComponent.ɵfac = function AddToCartComponent_Factory(t) { return new (t || AddToCartComponent)(ɵngcc0.ɵɵdirectiveInject(CartService)); };
AddToCartComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AddToCartComponent, selectors: [["add-to-cart"]], inputs: { custom: "custom", buttonText: "buttonText", buttonClass: "buttonClass", type: "type", position: "position", dropdown: "dropdown", item: "item", quantity: "quantity" }, outputs: { change: "change", added: "added" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 6, vars: 5, consts: [[1, "add-to-cart", 3, "ngClass"], ["class", "add-to-cart-component", 3, "ngClass", 4, "ngIf"], [1, "cart-button-container", 3, "click"], [4, "ngIf"], [1, "add-to-cart-component", 3, "ngClass"], [3, "type", "dropdown", "value", "valueChange"], ["type", "button", 3, "ngClass", "disabled"]], template: function AddToCartComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, AddToCartComponent_div_1_Template, 2, 4, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵlistener("click", function AddToCartComponent_Template_div_click_2_listener($event) { return ctx.addToCart($event); });
        ɵngcc0.ɵɵtemplate(3, AddToCartComponent_ng_container_3_Template, 3, 3, "ng-container", 3);
        ɵngcc0.ɵɵtemplate(4, AddToCartComponent_4_Template, 1, 0, undefined, 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, AddToCartComponent_div_5_Template, 2, 4, "div", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ctx.containerClass);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.editorPrecedence === "before" && ctx.hasEditor);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.custom);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.custom);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.editorPrecedence === "after" && ctx.hasEditor);
    } }, directives: function () { return [ɵngcc1.NgClass, ɵngcc1.NgIf, AddToCartEditorComponent]; }, encapsulation: 2 });
/** @nocollapse */
AddToCartComponent.ctorParameters = () => [
    { type: CartService, },
];
AddToCartComponent.propDecorators = {
    "custom": [{ type: Input },],
    "item": [{ type: Input },],
    "buttonText": [{ type: Input },],
    "buttonClass": [{ type: Input },],
    "type": [{ type: Input },],
    "position": [{ type: Input },],
    "dropdown": [{ type: Input },],
    "quantity": [{ type: Input },],
    "change": [{ type: Output },],
    "added": [{ type: Output },],
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AddToCartComponent, [{
        type: Component,
        args: [{
                selector: 'add-to-cart',
                // tslint:disable-line component-selector
                template: `<div class="add-to-cart" [ngClass]="containerClass">
  <div class="add-to-cart-component" [ngClass]="position"
       *ngIf="editorPrecedence === 'before' && hasEditor">
    <add-to-cart-editor [type]="type" [dropdown]="dropdown" [(value)]="editorQuantity"></add-to-cart-editor>
  </div>
  <div class="cart-button-container" (click)="addToCart($event)">
    <ng-container *ngIf="!custom">
      <button type="button" [ngClass]="buttonClass" [disabled]="!item">
        {{buttonText}}
      </button>
    </ng-container>
    <ng-content *ngIf="custom">
    </ng-content>
  </div>
  <div class="add-to-cart-component" [ngClass]="position"
       *ngIf="editorPrecedence === 'after' && hasEditor">
    <add-to-cart-editor [type]="type" [dropdown]="dropdown" [(value)]="editorQuantity"></add-to-cart-editor>
  </div>
</div>
`
            }]
    }], function () { return [{ type: CartService }]; }, { custom: [{
            type: Input
        }], buttonText: [{
            type: Input
        }], buttonClass: [{
            type: Input
        }], type: [{
            type: Input
        }], position: [{
            type: Input
        }], dropdown: [{
            type: Input
        }], change: [{
            type: Output
        }], added: [{
            type: Output
        }], item: [{
            type: Input
        }], quantity: [{
            type: Input
        }] }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Renders a button to initiate checkout of the cart.
 *
 * \@order 6
 * \@howToUse "With a custom button or projected content"
 * ```html
 * <cart-checkout [custom]="true">
 *    <button type="button" class="my-custom-class">Do checkout</button>
 * </cart-checkout>
 * ```
 *
 * \@howToUse "With different text and classes"
 * ```html
 * <cart-checkout [buttonText]="'Add item'" [buttonClass]="'my-custom-class'">
 * </cart-checkout>
 * ```
 *
 * \@howToUse "Using http in a protected endpoint"
 * ```html
 * <cart-checkout [service]="'http'" settings="settings">
 * </cart-checkout>
 * ```
 * ```typescript
 * export class MyComponent {
 *   settings: CheckoutHttpSettings = {
 *     method: 'POST',
 *     url: 'http://myapi.com/',
 *     options: { headers: { Authorization: 'Bearer my-auth-token' } }
 *   };
 * }
 * ```
 *
 * \@howToUse "Using the PayPal service"
 * ```html
 * <cart-checkout [service]="'paypal'" settings="settings">
 * </cart-checkout>
 * ```
 * ```typescript
 * export class MyComponent {
 *  settings: CheckoutPaypalSettings = {
 *    business: 'myaccount\@paypal.com',
 *    itemName: 'myMarketplaceAppCart',
 *    itemNumber: '1234',
 *    serviceName: 'MyBusiness',
 *    country: 'US'
 *  };
 * }
 * ```
 *
 * \@note {warning} This component captures clicks events bubbling from its projected content. Make sure the event keeps bubbling only when
 * you want the checkout operation to start.
 *
 * \@note {warning} When the `[service]` is set to `paypal` an actual PayPal button is rendered. None of the inputs `custom`, `buttonText`
 * or `buttonClass` have any effect.
 */
class CartCheckoutComponent {
    /**
     * @param {?} cartService
     * @param {?} httpClient
     * @param {?} locale
     */
    constructor(cartService, httpClient, locale) {
        this.cartService = cartService;
        this.httpClient = httpClient;
        this.locale = locale;
        this.empty = true;
        this.cost = 0;
        this.taxRate = 0;
        this.shipping = 0;
        this.currency = 'USD';
        this.paypalLocale = 'en';
        /**
         * If `false` displays a default button provided by the component. When set to `true` projects the contents of the component.
         */
        this.custom = false;
        /**
         * Changes the default text of the component's button.
         */
        this.buttonText = 'Checkout';
        /**
         * Changes the default text of the component's button.
         */
        this.buttonClass = 'cart-checkout-button';
        /**
         * Sets the type of service to be used when initiating the checkout.
         */
        this.service = 'log';
        /**
         * Depending on the type of the service you might need to add some configuration to it. This input allows you to change that
         * configuration.
         */
        this.settings = null;
        /**
         * Emits the result of the checkout operation. If the service is set to `'log'` it emits the entire cart object including tax rates and
         * shipping info. If is set to `'http'` it emits an `HttpResponse` object with body, headers, etc as it was received by the remote server.
         *
         * > When `[service]` is set to `'paypal'` this event is never emitted.
         */
        this.checkout = new EventEmitter();
        /**
         * When the `[service]` is set to `'http'` and the checkout operation fails the thrown error can be captured using this output.
         *
         * The emitted value is the complete `HttpErrorResponse` object returned by `HttpClient` so you can inspect other properties like status
         * codes, headers, messages, etc.
         */
        this.error = new EventEmitter();
        this.getLocaleCurrencyName = getLocaleCurrencyName;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.updateCart(true);
        this._serviceSubscription = this.cartService
            .onChange
            .subscribe((evt) => this.updateCart(evt.change === 'format'));
    }
    /**
     * @param {?} formatChange
     * @return {?}
     */
    updateCart(formatChange) {
        this.empty = this.cartService.isEmpty();
        this.cost = this.cartService.cost();
        this.taxRate = this.cartService.getTaxRate();
        this.shipping = this.cartService.getShipping();
        if (formatChange) {
            this.updateLocale();
        }
    }
    /**
     * @return {?}
     */
    updateLocale() {
        this.format = this.localeFormat ?
            parseLocaleFormat(this.localeFormat) : /** @type {?} */ (this.cartService.getLocaleFormat(true));
        const /** @type {?} */ loc = this.format.locale || this.locale;
        this.paypalLocale = loc.substring(0, 2);
        this.currency = this.format.currencyCode || this.getCurrency(loc);
    }
    /**
     * @param {?} locale
     * @return {?}
     */
    getCurrency(locale) {
        const /** @type {?} */ currencyCode = this.getLocaleCurrencyName(locale);
        if (!currencyCode) {
            return 'USD';
        }
        if (currencyCode.length === 3) {
            return currencyCode;
        }
        // Angular < 6 return "US Dollar" instead of "USD" so we have to hack the code using the currency pipe
        // You will also get USD on locales where you should get EUR so for those versions currencyCode must be used
        const /** @type {?} */ fmt = new CurrencyPipe(locale);
        const /** @type {?} */ val = fmt.transform(0, undefined, 'code', '1.0-0', locale);
        const /** @type {?} */ pre = val.startsWith('0');
        return val.substr(pre ? -3 : 0, 3);
    }
    /**
     * @return {?}
     */
    doCheckout() {
        let /** @type {?} */ cart = this.cartService.toObject();
        switch (this.service) {
            case 'log':
                console.log(cart);
                this.checkout.emit(cart);
                break;
            case 'http':
                if (!this.settings) {
                    throw new Error('Missing settings configuration');
                }
                const /** @type {?} */ verbs = ['POST', 'PUT', 'PATCH'];
                const { url, method = 'POST', options, body } = this.httpSettings;
                const /** @type {?} */ methodUpper = method.toUpperCase();
                if (verbs.indexOf(methodUpper) === -1) {
                    throw new Error(`Invalid http verb found in method setting. Expected one of ${verbs.join(' ')} and got ${method}`);
                }
                if (body) {
                    cart = typeof body === 'function' ? body(cart) : Object.assign({}, cart, body);
                }
                if (options && options.headers && options.headers.has('Content-Type')) {
                    const /** @type {?} */ contentType = options.headers.get('Content-Type');
                    if (contentType.startsWith('application/x-www-form-urlencoded')) {
                        cart = new HttpParams({ fromObject: cart });
                    }
                }
                this.httpClient
                    .request(new HttpRequest(methodUpper, url, cart, options))
                    .toPromise()
                    .then(response => {
                    this.checkout.emit(response);
                })
                    .catch(err => {
                    this.error.emit(err);
                });
                break;
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes['settings'] && changes['settings'].currentValue) {
            const /** @type {?} */ hasOwn = Object.prototype.hasOwnProperty;
            const /** @type {?} */ value = changes['settings'].currentValue;
            if (hasOwn.call(value, 'business')) {
                this.paypalSettings = changes['settings'].currentValue;
            }
            if (hasOwn.call(value, 'url')) {
                this.httpSettings = changes['settings'].currentValue;
            }
        }
        if (changes['localeFormat']) {
            this.updateLocale();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._serviceSubscription.unsubscribe();
    }
}
CartCheckoutComponent.ɵfac = function CartCheckoutComponent_Factory(t) { return new (t || CartCheckoutComponent)(ɵngcc0.ɵɵdirectiveInject(CartService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.HttpClient), ɵngcc0.ɵɵdirectiveInject(LOCALE_ID)); };
CartCheckoutComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CartCheckoutComponent, selectors: [["cart-checkout"]], inputs: { custom: "custom", buttonText: "buttonText", buttonClass: "buttonClass", service: "service", settings: "settings", localeFormat: "localeFormat" }, outputs: { checkout: "checkout", error: "error" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 2, vars: 2, consts: [[4, "ngIf"], [3, "ngClass", "disabled", "click", 4, "ngIf"], [3, "click", 4, "ngIf"], [3, "ngClass", "disabled", "click"], [3, "click"], ["action", "https://www.paypal.com/cgi-bin/webscr", "method", "post", "target", "_top"], ["type", "hidden", "name", "cmd", "value", "_xclick"], ["type", "hidden", "name", "business", 3, "value"], ["type", "hidden", "name", "lc", 3, "value"], ["type", "hidden", "name", "item_name", 3, "value", 4, "ngIf"], ["type", "hidden", "name", "item_number", 3, "value", 4, "ngIf"], ["type", "hidden", "name", "amount", 3, "value"], ["type", "hidden", "name", "currency_code", 3, "value"], ["type", "hidden", "name", "tax_rate", 3, "value"], ["type", "hidden", "name", "shipping", 3, "value"], ["type", "hidden", "name", "bn", 3, "value", 4, "ngIf"], ["type", "image", "src", "https://www.paypalobjects.com/en_US/i/btn/btn_buynowCC_LG.gif", "border", "0", "name", "submit", "alt", "PayPal - The safer, easier way to pay online!"], ["alt", "", "border", "0", "src", "https://www.paypalobjects.com/en_US/i/scr/pixel.gif", "width", "1", "height", "1"], ["type", "hidden", "name", "item_name", 3, "value"], ["type", "hidden", "name", "item_number", 3, "value"], ["type", "hidden", "name", "bn", 3, "value"]], template: function CartCheckoutComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, CartCheckoutComponent_ng_container_0_Template, 3, 2, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, CartCheckoutComponent_ng_container_1_Template, 14, 9, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.service === "log" || ctx.service === "http");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.service === "paypal");
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgClass], encapsulation: 2 });
/** @nocollapse */
CartCheckoutComponent.ctorParameters = () => [
    { type: CartService, },
    { type: HttpClient, },
    { type: undefined, decorators: [{ type: Inject, args: [LOCALE_ID,] },] },
];
CartCheckoutComponent.propDecorators = {
    "custom": [{ type: Input },],
    "buttonText": [{ type: Input },],
    "buttonClass": [{ type: Input },],
    "service": [{ type: Input },],
    "settings": [{ type: Input },],
    "localeFormat": [{ type: Input },],
    "checkout": [{ type: Output },],
    "error": [{ type: Output },],
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CartCheckoutComponent, [{
        type: Component,
        args: [{
                selector: 'cart-checkout',
                template: `<ng-container *ngIf="service === 'log' || service === 'http'">
  <button [ngClass]="buttonClass" [disabled]="empty" *ngIf="!custom" (click)="doCheckout()">
    {{buttonText}}
  </button>
  <span (click)="doCheckout()" *ngIf="custom">
    <ng-content>
    </ng-content>
  </span>
</ng-container>
<ng-container *ngIf="service === 'paypal'">
  <form action="https://www.paypal.com/cgi-bin/webscr" method="post" target="_top">
    <input type="hidden" name="cmd" value="_xclick"/>
    <input type="hidden" name="business" value="{{paypalSettings.business}}"/>
    <input type="hidden" name="lc" value="{{paypalLocale}}"/>
    <input type="hidden" name="item_name" value="{{paypalSettings.itemName}}" *ngIf="paypalSettings.itemName"/>
    <input type="hidden" name="item_number" value="{{paypalSettings.itemNumber}}" *ngIf="paypalSettings.itemNumber"/>
    <input type="hidden" name="amount" value="{{cost}}"/>
    <input type="hidden" name="currency_code" value="{{currency}}"/>
    <input type="hidden" name="tax_rate" value="{{taxRate}}"/>
    <input type="hidden" name="shipping" value="{{shipping}}"/>
    <input type="hidden" name="bn" value="{{paypalSettings.serviceName + '_BuyNow_WPS_' + paypalSettings.country}}"
           *ngIf="paypalSettings.serviceName && paypalSettings.country"/>
    <input type="image" src="https://www.paypalobjects.com/en_US/i/btn/btn_buynowCC_LG.gif" border="0" name="submit"
           alt="PayPal - The safer, easier way to pay online!"/>
    <img alt="" border="0" src="https://www.paypalobjects.com/en_US/i/scr/pixel.gif" width="1" height="1"/>
  </form>
</ng-container>
`
            }]
    }], function () { return [{ type: CartService }, { type: ɵngcc2.HttpClient }, { type: undefined, decorators: [{
                type: Inject,
                args: [LOCALE_ID]
            }] }]; }, { custom: [{
            type: Input
        }], buttonText: [{
            type: Input
        }], buttonClass: [{
            type: Input
        }], service: [{
            type: Input
        }], settings: [{
            type: Input
        }], checkout: [{
            type: Output
        }], error: [{
            type: Output
        }], localeFormat: [{
            type: Input
        }] }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Renders a summary of the contents of the cart.
 *
 * \@order 3
 * \@howToUse "With a different icon"
 * ```html
 * <cart-summary [icon]="'http://myapi/assets/icon.svg'"></cart-summary>
 * ```
 *
 * \@howToUse "Display different words when the cart changes"
 * ```html
 * <cart-summary [noItemsText]="'Zero items'" [oneItemText]="'Single item'" [manyItemsText]="'Exactly # items'"></cart-summary>
 * ```
 *
 * \@howToUse "Using always a number to display item count"
 * ```html
 * <cart-summary [noItemsText]="'# items'" [oneItemText]="'# items'" [manyItemsText]="'# items'"></cart-summary>
 * ```
 *
 * \@note {info} Inputs that allows you to customize text also accept the special character `#` to use numbers instead of words to
 * specify quantity, for example `'# bla'` will update to `'0 bla'` or `'1 bla'` when the number of items in the cart change.
 */
class CartSummaryComponent {
    /**
     * @param {?} cartService
     */
    constructor(cartService) {
        this.cartService = cartService;
        /**
         * The text to display when there are no items in the cart.
         */
        this.noItemsText = 'No items';
        /**
         * The text to display when there is only one item in the cart.
         */
        this.oneItemText = 'One item';
        /**
         * The text to display when there are several items in the cart.
         */
        this.manyItemsText = '# items';
        this.totalItems = 0;
        this.totalCost = 0;
    }
    /**
     * @return {?}
     */
    updateItemsText() {
        let /** @type {?} */ text = this.noItemsText;
        if (this.totalItems > 0) {
            text = this.totalItems === 1 ? this.oneItemText : this.manyItemsText;
        }
        this.itemsText = text.replace('#', this.totalItems.toString());
    }
    /**
     * @return {?}
     */
    updateComponent() {
        this.totalItems = this.cartService.itemCount();
        this.totalCost = !this.cartService.isEmpty() ? this.cartService.totalCost() : 0;
        if (!this.localeFormat) {
            this.format = /** @type {?} */ (this.cartService.getLocaleFormat(true));
        }
        this.updateItemsText();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.updateComponent();
        this._serviceSubscription = this.cartService.onChange.subscribe(() => {
            this.updateComponent();
        });
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes['localeFormat']) {
            this.format = this.localeFormat ?
                parseLocaleFormat(this.localeFormat) : /** @type {?} */ (this.cartService.getLocaleFormat(true));
        }
        if (changes['noItemsText'] || changes['oneItemText'] || changes['manyItemsText']) {
            this.updateItemsText();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._serviceSubscription.unsubscribe();
    }
}
CartSummaryComponent.ɵfac = function CartSummaryComponent_Factory(t) { return new (t || CartSummaryComponent)(ɵngcc0.ɵɵdirectiveInject(CartService)); };
CartSummaryComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CartSummaryComponent, selectors: [["cart-summary"]], inputs: { noItemsText: "noItemsText", oneItemText: "oneItemText", manyItemsText: "manyItemsText", icon: "icon", localeFormat: "localeFormat" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 10, vars: 16, consts: [[1, "cart-summary"], [1, "cart-summary-icon"], ["version", "1.1", "class", "summary-icon", "xmlns", "http://www.w3.org/2000/svg", 0, "xmlns", "xlink", "http://www.w3.org/1999/xlink", "viewBox", "0 0 32 32", 4, "ngIf"], ["class", "summary-icon", "alt", "cart-summary-icon", 3, "src", 4, "ngIf"], [1, "cart-summary-contents"], [1, "cart-summary-items"], [1, "cart-summary-cost"], ["version", "1.1", "xmlns", "http://www.w3.org/2000/svg", 0, "xmlns", "xlink", "http://www.w3.org/1999/xlink", "viewBox", "0 0 32 32", 1, "summary-icon"], ["stroke", "currentColor", "stroke-width", "2", "stroke-linecap", "round", "fill", "none", "stroke-linejoin", "round", "d", "M2,4 L6,4 L10,20 L21,20 L25,10 L8,10 M13,10 L13,20 M18,20 L18,10 M10,15 L23,15"], ["fill", "currentColor", "cx", "11", "cy", "24", "r", "2"], ["fill", "currentColor", "cx", "21", "cy", "24", "r", "2"], ["alt", "cart-summary-icon", 1, "summary-icon", 3, "src"]], template: function CartSummaryComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵtemplate(2, CartSummaryComponent__svg_svg_2_Template, 4, 0, "svg", 2);
        ɵngcc0.ɵɵtemplate(3, CartSummaryComponent_img_3_Template, 1, 1, "img", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 4);
        ɵngcc0.ɵɵelementStart(5, "div", 5);
        ɵngcc0.ɵɵtext(6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 6);
        ɵngcc0.ɵɵtext(8);
        ɵngcc0.ɵɵpipe(9, "currency");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.icon);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.icon);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.itemsText, " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBindV(9, 4, ɵngcc0.ɵɵpureFunction5(10, _c1, ctx.totalCost, ctx.format.currencyCode, ctx.format.display, ctx.format.digitsInfo, ctx.format.locale)), " ");
    } }, directives: [ɵngcc1.NgIf], pipes: [ɵngcc1.CurrencyPipe], encapsulation: 2 });
/** @nocollapse */
CartSummaryComponent.ctorParameters = () => [
    { type: CartService, },
];
CartSummaryComponent.propDecorators = {
    "icon": [{ type: Input },],
    "noItemsText": [{ type: Input },],
    "oneItemText": [{ type: Input },],
    "manyItemsText": [{ type: Input },],
    "localeFormat": [{ type: Input },],
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CartSummaryComponent, [{
        type: Component,
        args: [{
                selector: 'cart-summary',
                template: `<div class="cart-summary">
  <div class="cart-summary-icon">
    <svg *ngIf="!icon" version="1.1" class="summary-icon" xmlns="http://www.w3.org/2000/svg"
         xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 32 32">
      <path stroke="currentColor" stroke-width="2" stroke-linecap="round" fill="none" stroke-linejoin="round"
            d="M2,4 L6,4 L10,20 L21,20 L25,10 L8,10 M13,10 L13,20 M18,20 L18,10 M10,15 L23,15"/>
      <circle fill="currentColor" cx="11" cy="24" r="2"/>
      <circle fill="currentColor" cx="21" cy="24" r="2"/>
    </svg>
    <img *ngIf="icon" [src]="icon" class="summary-icon" alt="cart-summary-icon">
  </div>
  <div class="cart-summary-contents">
    <div class="cart-summary-items">
      {{itemsText}}
    </div>
    <div class="cart-summary-cost">
      {{totalCost | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
    </div>
  </div>
</div>
`
            }]
    }], function () { return [{ type: CartService }]; }, { noItemsText: [{
            type: Input
        }], oneItemText: [{
            type: Input
        }], manyItemsText: [{
            type: Input
        }], icon: [{
            type: Input
        }], localeFormat: [{
            type: Input
        }] }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * An editor to change the quantity of an item to add to the cart.
 * @ignore
 */
class AddToCartEditorComponent {
    constructor() {
        /**
         * The type of editor to display.
         */
        this.type = 'text';
        /**
         * The value to display in the editor
         */
        this.value = 1;
        /**
         * Emits a the value of the editor when the user changes it
         */
        this.valueChange = new EventEmitter();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    changeValue(value) {
        const /** @type {?} */ val = parseFloat(value);
        this.valueChange.emit(Number.isNaN(val) ? 1 : val);
    }
}
AddToCartEditorComponent.ɵfac = function AddToCartEditorComponent_Factory(t) { return new (t || AddToCartEditorComponent)(); };
AddToCartEditorComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AddToCartEditorComponent, selectors: [["add-to-cart-editor"]], inputs: { type: "type", value: "value", dropdown: "dropdown" }, outputs: { valueChange: "valueChange" }, decls: 2, vars: 2, consts: [[4, "ngIf"], [1, "add-to-cart-input", 3, "change"], ["selectAmount", ""], [3, "selected", "value", 4, "ngFor", "ngForOf"], [3, "selected", "value"], [1, "add-to-cart-input", 3, "type", "value", "change"], ["inputAmount", ""]], template: function AddToCartEditorComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, AddToCartEditorComponent_ng_container_0_Template, 4, 1, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, AddToCartEditorComponent_ng_container_1_Template, 3, 2, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.type === "dropdown");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.type === "number" || ctx.type === "text");
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgForOf], encapsulation: 2 });
/** @nocollapse */
AddToCartEditorComponent.ctorParameters = () => [];
AddToCartEditorComponent.propDecorators = {
    "type": [{ type: Input },],
    "dropdown": [{ type: Input },],
    "value": [{ type: Input },],
    "valueChange": [{ type: Output },],
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AddToCartEditorComponent, [{
        type: Component,
        args: [{
                selector: 'add-to-cart-editor',
                // tslint:disable-line component-selector
                template: `<ng-container *ngIf="type === 'dropdown'">
  <select class="add-to-cart-input" #selectAmount (change)="changeValue(selectAmount.value)">
    <option [selected]="item.value === value" *ngFor="let item of dropdown" [value]="item.value">{{item.label}}</option>
  </select>
</ng-container>
<ng-container *ngIf="type === 'number' || type === 'text'">
  <input class="add-to-cart-input" #inputAmount [type]="type" (change)="changeValue(inputAmount.value)" [value]="value">
</ng-container>
`
            }]
    }], function () { return []; }, { type: [{
            type: Input
        }], value: [{
            type: Input
        }], valueChange: [{
            type: Output
        }], dropdown: [{
            type: Input
        }] }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Renders a view of the cart.
 *
 * \@order 2
 * \@howToUse "Using responsive layout"
 * ```html
 * <cart-view [display]="'responsive'">
 * </cart-view>
 * ```
 *
 * \@howToUse "No images and using scrollbars on small screens"
 * ```html
 * <cart-view [images]="false" [display]="'responsive-table'">
 * </cart-view>
 * ```
 *
 * \@howToUse "Using different text for headers"
 * ```html
 * <cart-view [emptyText]="headers.empty" [nameHeaderText]="headers.name" [quantityHeaderText]="headers.quantity"
 *  [priceHeaderText]="headers.quantity" [totalHeaderText]="headers.total" [taxFooterText]="footers.tax"
 *  [shippingFooterText]="footers.shipping" [totalFooterText]="footers.total"
 * >
 * </cart-view>
 * ```
 * ```typescript
 * export class MyComponent {
 *   headers = {
 *     empty: 'No items. Add some to the cart',
 *     name: 'Description',
 *     quantity: 'Amount',
 *     price: 'Cost',
 *     total: 'Total x item',
 *   }
 *   footers = {
 *     tax: 'Tax rate',
 *     shipping: 'Shipping cost',
 *     total: 'Total cost'
 *   }
 * }
 * ```
 *
 * \@howToUse "Change the default empty cart content"
 * ```html
 * <cart-view [customEmptyContent]="true">
 *   <div class="my-empty-cart-view">
 *       <span style="font-size: 36px;" class="glyphicon glyphicon-shopping-cart" aria-hidden="true"></span>
 *       Your cart is empty
 *   </div>
 * </cart-view>
 * ```
 */
class CartViewComponent {
    /**
     * @param {?} cartService
     */
    constructor(cartService) {
        this.cartService = cartService;
        /**
         * Changes the appearance how the cart view displays in different screen sizes
         */
        this.display = 'responsive-table';
        /**
         * Whether to include images in the cart or not.
         */
        this.images = true;
        /**
         * The text to show when the cart has no items in it.
         */
        this.emptyText = 'Your cart is empty';
        /**
         * When set to `true` and the cart is empty displays the projected content of the component as the empty content.
         */
        this.customEmptyContent = false;
        /**
         * The text to display in the header of the name column.
         */
        this.nameHeaderText = 'Name';
        /**
         * The text to display in the header of the quantity column.
         */
        this.quantityHeaderText = 'Quantity';
        /**
         * The text to display in the header of the price column.
         */
        this.priceHeaderText = 'Price';
        /**
         * The text to display in the header of the total per item column.
         */
        this.totalHeaderText = 'Total';
        /**
         * The text to display in the tax section of the footer.
         */
        this.taxFooterText = 'Tax';
        /**
         * The text to display in the shipping section of the footer.
         */
        this.shippingFooterText = 'Shipping';
        /**
         * The text to display in the total section of the footer.
         */
        this.totalFooterText = 'Total';
        this.empty = true;
        this.taxRate = 0;
        this.tax = 0;
        this.shipping = 0;
        this.cost = 0;
    }
    /**
     * @return {?}
     */
    update() {
        this.empty = this.cartService.isEmpty();
        this.items = this.cartService.getItems();
        this.taxRate = this.cartService.getTaxRate() / 100;
        this.tax = this.cartService.getTax();
        this.shipping = this.cartService.getShipping();
        this.cost = this.cartService.totalCost();
        if (!this.localeFormat) {
            this.format = /** @type {?} */ (this.cartService.getLocaleFormat(true));
        }
    }
    /**
     * @param {?} item
     * @return {?}
     */
    increase(item) {
        item.setQuantity(item.getQuantity() + 1);
        this.cartService.addItem(item);
    }
    /**
     * @param {?} item
     * @return {?}
     */
    decrease(item) {
        if (item.getQuantity() > 1) {
            item.setQuantity(item.getQuantity() - 1);
            this.cartService.addItem(item);
        }
        else {
            this.cartService.removeItem(item.getId());
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.update();
        this._serviceSubscription = this.cartService.onChange.subscribe(() => {
            this.update();
        });
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes['localeFormat']) {
            this.format = this.localeFormat ?
                parseLocaleFormat(this.localeFormat) : /** @type {?} */ (this.cartService.getLocaleFormat(true));
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._serviceSubscription.unsubscribe();
    }
}
CartViewComponent.ɵfac = function CartViewComponent_Factory(t) { return new (t || CartViewComponent)(ɵngcc0.ɵɵdirectiveInject(CartService)); };
CartViewComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CartViewComponent, selectors: [["cart-view"]], inputs: { display: "display", images: "images", emptyText: "emptyText", customEmptyContent: "customEmptyContent", nameHeaderText: "nameHeaderText", quantityHeaderText: "quantityHeaderText", priceHeaderText: "priceHeaderText", totalHeaderText: "totalHeaderText", taxFooterText: "taxFooterText", shippingFooterText: "shippingFooterText", totalFooterText: "totalFooterText", localeFormat: "localeFormat" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 2, vars: 2, consts: [[4, "ngIf"], ["class", "cart-view-empty", 4, "ngIf"], [1, "cart-view-empty"], ["class", "cart-view", 3, "ngClass", 4, "ngIf"], [1, "cart-view", 3, "ngClass"], [1, "cart-list-view", 3, "ngClass"], [1, "cart-list-header"], [1, "cart-list-header-value", "cart-list-image-header"], [1, "cart-list-header-value", "cart-list-name-header"], [1, "cart-list-header-value", "cart-list-quantity-header"], [1, "cart-list-header-value", "cart-list-price-header"], [1, "cart-list-header-value", "cart-list-total-header"], ["class", "cart-list-item", 4, "ngFor", "ngForOf"], [1, "cart-list-footer"], [1, "cart-list-summary"], ["class", "cart-empty-summary", 4, "ngIf"], [1, "cart-empty-summary"], [1, "cart-tax-summary"], [1, "cart-shipping-summary"], [1, "cart-total-summary"], [1, "cart-list-item"], [1, "cart-list-value", "cart-list-image-value", 3, "ngStyle"], [1, "cart-list-value", "cart-list-name-value"], [1, "cart-list-value", "cart-list-quantity-value"], ["type", "button", 1, "cart-increase-button", 3, "click"], [1, "cart-list-quantity-content"], ["type", "button", 1, "cart-decrease-button", 3, "click"], [1, "cart-list-value", "cart-list-price-value"], [1, "cart-list-value", "cart-list-total-value"], [1, "cart-view", "table-display"], ["class", "cart-list-header-value cart-list-image-header", 4, "ngIf"], [4, "ngFor", "ngForOf"], ["class", "cart-list-value cart-list-image-value", 3, "ngStyle", 4, "ngIf"]], template: function CartViewComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, CartViewComponent_ng_container_0_Template, 3, 2, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, CartViewComponent_ng_container_1_Template, 3, 2, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.empty);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.empty);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc1.NgForOf, ɵngcc1.NgStyle], pipes: [ɵngcc1.PercentPipe, ɵngcc1.CurrencyPipe], encapsulation: 2 });
/** @nocollapse */
CartViewComponent.ctorParameters = () => [
    { type: CartService, },
];
CartViewComponent.propDecorators = {
    "display": [{ type: Input },],
    "images": [{ type: Input },],
    "emptyText": [{ type: Input },],
    "customEmptyContent": [{ type: Input },],
    "nameHeaderText": [{ type: Input },],
    "quantityHeaderText": [{ type: Input },],
    "priceHeaderText": [{ type: Input },],
    "totalHeaderText": [{ type: Input },],
    "taxFooterText": [{ type: Input },],
    "shippingFooterText": [{ type: Input },],
    "totalFooterText": [{ type: Input },],
    "localeFormat": [{ type: Input },],
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CartViewComponent, [{
        type: Component,
        args: [{
                selector: 'cart-view',
                template: `<ng-container *ngIf="empty">
  <div *ngIf="!customEmptyContent" class="cart-view-empty">
    {{emptyText}}
  </div>
  <ng-content *ngIf="customEmptyContent"></ng-content>
</ng-container>
<ng-container *ngIf="!empty">
  <ng-container *ngIf="display !== 'table'">
    <div *ngIf="!empty" class="cart-view" [ngClass]="display + '-display'">
      <div class="cart-list-view" [ngClass]="{'no-images': !images}">
        <div class="cart-list-header">
          <div class="cart-list-header-value cart-list-image-header">
          </div>
          <div class="cart-list-header-value cart-list-name-header">
            {{nameHeaderText}}
          </div>
          <div class="cart-list-header-value cart-list-quantity-header">
            {{quantityHeaderText}}
          </div>
          <div class="cart-list-header-value cart-list-price-header">
            {{priceHeaderText}}
          </div>
          <div class="cart-list-header-value cart-list-total-header">
            {{totalHeaderText}}
          </div>
        </div>
        <div *ngFor="let cartItem of items" class="cart-list-item">
          <div class="cart-list-value cart-list-image-value"
               [ngStyle]="{'background-image': 'url(' + cartItem.getImage() + ')'}">
          </div>
          <div class="cart-list-value cart-list-name-value">
            {{cartItem.getName()}}
          </div>
          <div class="cart-list-value cart-list-quantity-value">
            <button type="button" class="cart-increase-button" (click)="increase(cartItem)">
              +
            </button>
            <span class="cart-list-quantity-content">{{cartItem.getQuantity()}}</span>
            <button type="button" class="cart-decrease-button" (click)="decrease(cartItem)">
              -
            </button>
          </div>
          <div class="cart-list-value cart-list-price-value">
            {{cartItem.getPrice() | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
          </div>
          <div class="cart-list-value cart-list-total-value">
            {{cartItem.total() | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
          </div>
        </div>
        <div class="cart-list-footer">
          <div class="cart-list-summary">
            <div class="cart-empty-summary" *ngIf="images"></div>
            <div class="cart-empty-summary"></div>
            <div class="cart-empty-summary"></div>
            <div class="cart-empty-summary"></div>
            <div class="cart-tax-summary">
              {{taxFooterText}}: ({{taxRate | percent:format.digitsInfo:format.locale}})
              {{tax | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
            </div>
          </div>
          <div class="cart-list-summary">
            <div class="cart-empty-summary" *ngIf="images"></div>
            <div class="cart-empty-summary"></div>
            <div class="cart-empty-summary"></div>
            <div class="cart-empty-summary"></div>
            <div class="cart-shipping-summary">
              {{shippingFooterText}}:
              {{shipping | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
            </div>
          </div>
          <div class="cart-list-summary">
            <div class="cart-empty-summary" *ngIf="images"></div>
            <div class="cart-empty-summary"></div>
            <div class="cart-empty-summary"></div>
            <div class="cart-empty-summary"></div>
            <div class="cart-total-summary">
              {{totalFooterText}}:
              {{cost | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
            </div>
          </div>
        </div>
      </div>
    </div>
  </ng-container>
  <ng-container *ngIf="display === 'table'">
    <table class="cart-view table-display">
      <thead class="cart-list-header">
      <tr>
        <th class="cart-list-header-value cart-list-image-header" *ngIf="images"></th>
        <th class="cart-list-header-value cart-list-name-header">{{nameHeaderText}}</th>
        <th class="cart-list-header-value cart-list-quantity-header">{{quantityHeaderText}}</th>
        <th class="cart-list-header-value cart-list-price-header">{{priceHeaderText}}</th>
        <th class="cart-list-header-value cart-list-total-header">{{totalHeaderText}}</th>
      </tr>
      </thead>
      <tbody>
      <tr *ngFor="let cartItem of items">
        <td class="cart-list-value cart-list-image-value" *ngIf="images"
            [ngStyle]="{'background-image': 'url(' + cartItem.getImage() + ')'}">
        </td>
        <td class="cart-list-value cart-list-name-value">{{cartItem.getName()}}</td>
        <td class="cart-list-value cart-list-quantity-value">
          <button type="button" class="cart-increase-button" (click)="increase(cartItem)">
            +
          </button>
          <span class="cart-list-quantity-content">{{cartItem.getQuantity()}}</span>
          <button type="button" class="cart-decrease-button" (click)="decrease(cartItem)">
            -
          </button>
        </td>
        <td class="cart-list-value cart-list-price-value">
          {{cartItem.getPrice() | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
        </td>
        <td class="cart-list-value cart-list-total-value">
          {{cartItem.total() | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
        </td>
      </tr>
      </tbody>
      <tfoot class="cart-list-footer">
      <tr class="cart-list-summary">
        <td [attr.colspan]="images ? 4 : 3" class="cart-empty-summary"></td>
        <td class="cart-tax-summary">
          {{taxFooterText}}: ({{taxRate | percent:format.digitsInfo:format.locale}})
          {{tax | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
        </td>
      </tr>
      <tr class="cart-list-summary">
        <td [attr.colspan]="images ? 4 : 3" class="cart-empty-summary"></td>
        <td class="cart-shipping-summary">
          {{shippingFooterText}}:
          {{shipping | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
        </td>
      </tr>
      <tr class="cart-list-summary">
        <td [attr.colspan]="images ? 4 : 3" class="cart-empty-summary"></td>
        <td class="cart-total-summary">
          {{totalFooterText}}:
          {{cost | currency:format.currencyCode:format.display:format.digitsInfo:format.locale}}
        </td>
      </tr>
      </tfoot>
    </table>
  </ng-container>
</ng-container>
`
            }]
    }], function () { return [{ type: CartService }]; }, { display: [{
            type: Input
        }], images: [{
            type: Input
        }], emptyText: [{
            type: Input
        }], customEmptyContent: [{
            type: Input
        }], nameHeaderText: [{
            type: Input
        }], quantityHeaderText: [{
            type: Input
        }], priceHeaderText: [{
            type: Input
        }], totalHeaderText: [{
            type: Input
        }], taxFooterText: [{
            type: Input
        }], shippingFooterText: [{
            type: Input
        }], totalFooterText: [{
            type: Input
        }], localeFormat: [{
            type: Input
        }] }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * The default implementation of a showcase item.
 *
 * \@note {info} This component is only intended to be used as a template for you to create your own components either by pure css
 * customization or by providing an actually working sample to serve as a guide for more complex cases.
 *
 * \@order 5
 */
class CartShowcaseItemComponent {
}
CartShowcaseItemComponent.ɵfac = function CartShowcaseItemComponent_Factory(t) { return new (t || CartShowcaseItemComponent)(); };
CartShowcaseItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CartShowcaseItemComponent, selectors: [["cart-showcase-item"]], decls: 7, vars: 8, consts: [[1, "showcase-item", "default-sc-item", 3, "ngStyle"], [1, "default-sc-description"], [1, "default-sc-name"], [1, "default-sc-price"]], template: function CartShowcaseItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵpipe(6, "cartCurrency");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(6, _c2, "url(" + ctx.item.getImage() + ")"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.item.getName(), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind2(6, 3, ctx.item.getPrice(), ctx.format), " ");
    } }, directives: function () { return [ɵngcc1.NgStyle]; }, pipes: function () { return [CartCurrencyPipe]; }, encapsulation: 2 });
/** @nocollapse */
CartShowcaseItemComponent.ctorParameters = () => [];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CartShowcaseItemComponent, [{
        type: Component,
        args: [{
                selector: 'cart-showcase-item',
                template: `<div class="showcase-item default-sc-item" [ngStyle]="{'background-image': 'url(' + item.getImage() + ')'}">
  <div class="default-sc-description">
    <div class="default-sc-name">
      {{item.getName()}}
    </div>
    <div class="default-sc-price">
      {{item.getPrice() | cartCurrency:format}}
    </div>
  </div>
</div>
`
            }]
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Renders items arranged in columns using a dynamic component for the item. Useful for getting started with e-commerce applications.
 *
 * \@order 4
 * \@howToUse "Using wider items"
 * ```html
 * <cart-showcase [items]="items" [aspectRatio]="'2:1'">
 * </cart-showcase>
 * ```
 *
 * \@howToUse "Using four columns in all screen sizes bigger than 768px"
 * ```html
 * <cart-showcase [items]="items" [mCols]="4" [lCols]="4">
 * </cart-showcase>
 * ```
 *
 * \@howToUse "Using a different item component"
 * ```html
 * <!-- my-component.html -->
 * <cart-showcase [items]="items" [itemComponent]="itemComponent">
 * </cart-showcase>
 * ```
 * ```typescript
 * // my-component.ts
 * export class MyComponent {
 *   itemComponent = MyCustomItemComponent;
 * }
 *
 * // my-custom-item-component.ts
 * \@Component({
 *  selector: 'my-custom-item-component',
 *  template: '<div class="item-class">{{item.getName()}}</div>'
 * })
 * export class MyCustomItemComponent implements ShowcaseItem  {
 *   item: CartItem;
 * }
 *
 * // app.module.ts
 * \@NgModule({
 *   // .....
 *   entryComponents: [MyCustomItemComponent],
 * })
 * export class AppModule {
 * }
 * ```
 *
 * \@note {warning} If you change the `[columns]` input you must also change the sass variable that controls the component grid and
 * vice-versa. A similar procedure is required to create aspect ratios with values greater than four eg: `'1:5'`. Check the styling guide
 * for more information.
 *
 * \@note {danger} The aspect ratio is the width/height proportion of the items therefore a ratio of `'2:2'` is equivalent to `'1:1'`.
 * Redundant ratios like these are removed from the source so don't try to use them.
 *
 */
class CartShowcaseComponent {
    /**
     * @param {?} cartService
     */
    constructor(cartService) {
        this.cartService = cartService;
        this.xsClass = 'sc-container-xs-12';
        this.sClass = 'sc-container-s-6';
        this.mClass = 'sc-container-m-4';
        this.lClass = 'sc-container-l-4';
        this.xlClass = 'sc-container-xl-3';
        this.ratioClass = 'sc-ratio-1-1';
        /**
         * The number of columns to display when the screen size matches phone devices.
         */
        this.xsCols = 1;
        /**
         * The number of columns to display when the screen matches tablet devices.
         */
        this.sCols = 2;
        /**
         * The number of columns to display when the screen matches desktop devices.
         */
        this.mCols = 3;
        /**
         * The number of columns to display when the screen matches large desktop devices.
         */
        this.lCols = 3;
        /**
         * The number of columns to display when the screen matches extra large desktop devices.
         */
        this.xlCols = 4;
        /**
         * The number of columns in the grid.
         * Only update this value if you changed the columns sass variable in the library styles following the Styling guide.
         */
        this.columns = 12;
        /**
         * The component to render for each item. This type means any component that implements the interface `ShowcaseItem`.
         */
        this.itemComponent = CartShowcaseItemComponent;
        /**
         * The aspect ratio of the container of the items. A value of `1:1` means square items, `2:1` means two times wider, `1:2` two times
         * taller and so on.
         */
        this.aspectRatio = '1:1';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    getColumnSize(value) {
        return Math.floor(this.columns / value);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        const /** @type {?} */ columnProps = ['xsCols', 'sCols', 'mCols', 'lCols', 'xlCols'];
        const /** @type {?} */ classPrefix = ['xs', 's', 'm', 'l', 'xl'];
        for (let /** @type {?} */ i = 0; i < columnProps.length; i++) {
            const /** @type {?} */ prop = columnProps[i];
            const /** @type {?} */ colChanges = changes[prop];
            if (changes['columns'] || colChanges) {
                const /** @type {?} */ prefix = classPrefix[i];
                const /** @type {?} */ size = this.getColumnSize(this[prop]);
                this[`${prefix}Class`] = `sc-container-${prefix}-${size}`;
            }
        }
        if (changes['aspectRatio']) {
            const /** @type {?} */ newRatio = changes['aspectRatio'].currentValue;
            const /** @type {?} */ values = newRatio.split(':');
            if (values.length === 2) {
                this.ratioClass = `sc-ratio-${values[0]}-${values[1]}`;
            }
        }
        if (changes['localeFormat']) {
            this.format = this.localeFormat || /** @type {?} */ (this.cartService.getLocaleFormat());
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.format = this.localeFormat || /** @type {?} */ (this.cartService.getLocaleFormat());
        this._serviceSubscription = this.cartService.onChange.subscribe((evt) => {
            if (evt.change === 'format' && !this.localeFormat) {
                this.format = /** @type {?} */ (this.cartService.getLocaleFormat());
            }
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._serviceSubscription.unsubscribe();
    }
}
CartShowcaseComponent.ɵfac = function CartShowcaseComponent_Factory(t) { return new (t || CartShowcaseComponent)(ɵngcc0.ɵɵdirectiveInject(CartService)); };
CartShowcaseComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CartShowcaseComponent, selectors: [["cart-showcase"]], inputs: { xsCols: "xsCols", sCols: "sCols", mCols: "mCols", lCols: "lCols", xlCols: "xlCols", columns: "columns", itemComponent: "itemComponent", aspectRatio: "aspectRatio", items: "items", injector: "injector", moduleFactory: "moduleFactory", localeFormat: "localeFormat" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 2, vars: 1, consts: [[1, "cart-showcase"], ["class", "sc-item-container", 3, "ngClass", 4, "ngFor", "ngForOf"], [1, "sc-item-container", 3, "ngClass"], [1, "sc-item-wrapper"], [4, "cartShowcaseOutlet", "cartShowcaseOutletItem", "cartShowcaseOutletFormat", "cartShowcaseOutletInjector", "cartShowcaseOutletNgModuleFactory"]], template: function CartShowcaseComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, CartShowcaseComponent_div_1_Template, 3, 13, "div", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.items);
    } }, directives: function () { return [ɵngcc1.NgForOf, ɵngcc1.NgClass, ShowcaseOutletDirective]; }, encapsulation: 2 });
/** @nocollapse */
CartShowcaseComponent.ctorParameters = () => [
    { type: CartService, },
];
CartShowcaseComponent.propDecorators = {
    "xsCols": [{ type: Input },],
    "sCols": [{ type: Input },],
    "mCols": [{ type: Input },],
    "lCols": [{ type: Input },],
    "xlCols": [{ type: Input },],
    "columns": [{ type: Input },],
    "items": [{ type: Input },],
    "itemComponent": [{ type: Input },],
    "injector": [{ type: Input },],
    "moduleFactory": [{ type: Input },],
    "aspectRatio": [{ type: Input },],
    "localeFormat": [{ type: Input },],
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CartShowcaseComponent, [{
        type: Component,
        args: [{
                selector: 'cart-showcase',
                template: `<div class="cart-showcase">
  <div class="sc-item-container" *ngFor="let carItem of items"
       [ngClass]="[xsClass, sClass, mClass, lClass, xlClass, ratioClass]">
    <div class="sc-item-wrapper">
      <ng-container
        *cartShowcaseOutlet="itemComponent;item:carItem;format:format;injector:injector;ngModuleFactory:moduleFactory">
      </ng-container>
    </div>
  </div>
</div>
`
            }]
    }], function () { return [{ type: CartService }]; }, { xsCols: [{
            type: Input
        }], sCols: [{
            type: Input
        }], mCols: [{
            type: Input
        }], lCols: [{
            type: Input
        }], xlCols: [{
            type: Input
        }], columns: [{
            type: Input
        }], itemComponent: [{
            type: Input
        }], aspectRatio: [{
            type: Input
        }], items: [{
            type: Input
        }], injector: [{
            type: Input
        }], moduleFactory: [{
            type: Input
        }], localeFormat: [{
            type: Input
        }] }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * A directive to create dynamic item components for the showcase component
 * @ignore
 */
class ShowcaseOutletDirective {
    /**
     * @param {?} viewContainerRef
     */
    constructor(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
        this._componentRef = null;
        this._moduleRef = null;
    }
    /**
     * @return {?}
     */
    cleanModule() {
        if (this._moduleRef) {
            this._moduleRef.destroy();
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        const /** @type {?} */ templateChange = Object.keys(changes).length !== 1 || !changes['cartShowcaseOutletFormat'];
        if (templateChange) {
            this.viewContainerRef.clear();
            this._componentRef = null;
            if (this.cartShowcaseOutlet) {
                const /** @type {?} */ elInjector = this.cartShowcaseOutletInjector || this.viewContainerRef.parentInjector;
                if (changes['cartShowcaseOutletNgModuleFactory']) {
                    this.cleanModule();
                    if (this.cartShowcaseOutletNgModuleFactory) {
                        const /** @type {?} */ parentModule = elInjector.get(NgModuleRef);
                        this._moduleRef = this.cartShowcaseOutletNgModuleFactory.create(parentModule.injector);
                    }
                    else {
                        this._moduleRef = null;
                    }
                }
                const /** @type {?} */ componentFactoryResolver = this._moduleRef ? this._moduleRef.componentFactoryResolver :
                    elInjector.get(ComponentFactoryResolver);
                const /** @type {?} */ componentFactory = componentFactoryResolver.resolveComponentFactory(this.cartShowcaseOutlet);
                this._componentRef = this.viewContainerRef.createComponent(componentFactory, this.viewContainerRef.length, elInjector);
                const /** @type {?} */ instance = this._componentRef.instance;
                instance.item = this.cartShowcaseOutletItem;
                instance.format = this.cartShowcaseOutletFormat;
            }
        }
        else {
            if (this._componentRef) {
                this._componentRef.instance.format = this.cartShowcaseOutletFormat;
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.cleanModule();
    }
}
ShowcaseOutletDirective.ɵfac = function ShowcaseOutletDirective_Factory(t) { return new (t || ShowcaseOutletDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
ShowcaseOutletDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ShowcaseOutletDirective, selectors: [["", "cartShowcaseOutlet", ""]], inputs: { cartShowcaseOutlet: "cartShowcaseOutlet", cartShowcaseOutletInjector: "cartShowcaseOutletInjector", cartShowcaseOutletNgModuleFactory: "cartShowcaseOutletNgModuleFactory", cartShowcaseOutletItem: "cartShowcaseOutletItem", cartShowcaseOutletFormat: "cartShowcaseOutletFormat" }, features: [ɵngcc0.ɵɵNgOnChangesFeature] });
/** @nocollapse */
ShowcaseOutletDirective.ctorParameters = () => [
    { type: ViewContainerRef, },
];
ShowcaseOutletDirective.propDecorators = {
    "cartShowcaseOutlet": [{ type: Input },],
    "cartShowcaseOutletInjector": [{ type: Input },],
    "cartShowcaseOutletNgModuleFactory": [{ type: Input },],
    "cartShowcaseOutletItem": [{ type: Input },],
    "cartShowcaseOutletFormat": [{ type: Input },],
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ShowcaseOutletDirective, [{
        type: Directive,
        args: [{
                selector: '[cartShowcaseOutlet]'
            }]
    }], function () { return [{ type: ɵngcc0.ViewContainerRef }]; }, { cartShowcaseOutlet: [{
            type: Input
        }], cartShowcaseOutletInjector: [{
            type: Input
        }], cartShowcaseOutletNgModuleFactory: [{
            type: Input
        }], cartShowcaseOutletItem: [{
            type: Input
        }], cartShowcaseOutletFormat: [{
            type: Input
        }] }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * An injection token to resolve the class used to create CartItem instances
 */
const CART_ITEM_CLASS = new InjectionToken('CartItemClass');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * An injection token to resolve the configuration of the cart service
 */
const CART_SERVICE_CONFIGURATION = new InjectionToken('CartServiceConfiguration');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * An injection token to store the service type entered in the `forRoot` static function. Is used to prevent errors when compiling with AOT.
 * @ignore
 *
 * \@note {info} You can safely ignore this token if you are using custom cart services.
 */
const CART_SERVICE_TYPE = new InjectionToken('CartServiceType');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// unsupported: template constraints.
/**
 * An implementation of the CartService using an in-memory array to store items
 * \@order 2
 * @template T
 */
class MemoryCartService extends CartService {
    constructor() {
        super(...arguments);
        this._items = [];
        this._taxRate = 0;
        this._shipping = 0;
    }
    /**
     * @param {?} item
     * @return {?}
     */
    _addItem(item) {
        const /** @type {?} */ foundIdx = this._items.findIndex(i => i.getId() === item.getId());
        if (foundIdx === -1) {
            this._items.push(item);
        }
        else {
            this._items[foundIdx] = item;
        }
        this.onItemAdded.emit(item);
        this.onItemsChanged.emit(this._items.length);
        this.onChange.emit({ change: 'items', value: this.getItems() });
    }
    /**
     * @param {?} id
     * @return {?}
     */
    _removeItem(id) {
        const /** @type {?} */ idx = this._items.findIndex(i => i.getId() === id);
        if (idx !== -1) {
            const /** @type {?} */ removed = this._items.splice(idx, 1);
            this.onItemRemoved.emit(removed[0]);
            this.onItemsChanged.emit(this._items.length);
            this.onChange.emit({ change: 'items', value: this.getItems() });
        }
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getItem(id) {
        return this._items.find(i => i.getId() === id);
    }
    /**
     * @return {?}
     */
    getItems() {
        return this._items.slice();
    }
    /**
     * @return {?}
     */
    itemCount() {
        return this._items.length;
    }
    /**
     * @return {?}
     */
    entries() {
        return this._items.reduce((curr, i) => (curr + i.getQuantity()), 0);
    }
    /**
     * @param {?} item
     * @return {?}
     */
    addItem(item) {
        this._addItem(item);
    }
    /**
     * @param {?} id
     * @return {?}
     */
    removeItem(id) {
        this._removeItem(id);
    }
    /**
     * @return {?}
     */
    cost() {
        return this._items.reduce((curr, i) => (curr + i.getPrice() * i.getQuantity()), 0);
    }
    /**
     * @return {?}
     */
    clear() {
        this._items = [];
        this.onItemsChanged.emit(this._items.length);
        this.onChange.emit({ change: 'items', value: this.getItems() });
    }
    /**
     * @return {?}
     */
    getShipping() {
        return this._shipping;
    }
    /**
     * @param {?} shipping
     * @return {?}
     */
    setShipping(shipping) {
        this._shipping = shipping;
        this.onShippingChange.emit(this._shipping);
        this.onChange.emit({ change: 'shipping', value: this._shipping });
    }
    /**
     * @return {?}
     */
    getTaxRate() {
        return this._taxRate;
    }
    /**
     * @param {?} taxRate
     * @return {?}
     */
    setTaxRate(taxRate) {
        this._taxRate = taxRate;
        this.onTaxChange.emit(this._taxRate);
        this.onChange.emit({ change: 'taxRate', value: this._taxRate });
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return this._items.length === 0;
    }
}
MemoryCartService.ɵfac = function MemoryCartService_Factory(t) { return ɵMemoryCartService_BaseFactory(t || MemoryCartService); };
MemoryCartService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: MemoryCartService, factory: MemoryCartService.ɵfac });
/** @nocollapse */
MemoryCartService.ctorParameters = () => [];
const ɵMemoryCartService_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(MemoryCartService);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MemoryCartService, [{
        type: Injectable
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// unsupported: template constraints.
/**
 * The base class for all CartService implementations that use the Storage interface of the Web Storage API like LocalStorage and
 * SessionStorage.
 * \@service
 *
 * \@order 3
 * @abstract
 * @template T
 */
class BrowserStorageCartService extends MemoryCartService {
    /**
     * @param {?} itemClass
     * @param {?} configuration
     */
    constructor(itemClass, configuration) {
        super() /* istanbul ignore next */;
        this.storageKey = configuration && configuration.storageKey ? configuration.storageKey : 'NgShoppingCart';
        this.clearOnError = configuration && configuration.clearOnError !== undefined ? configuration.clearOnError : true;
        this.itemClass = itemClass;
    }
    /**
     * @param {?} error
     * @return {?}
     */
    resetStorage(error) {
        if (this.clearOnError || !error) {
            this.setTaxRate(0);
            this.setShipping(0);
            this.clear();
            this.save();
        }
        else {
            if (typeof error === 'string') {
                throw new Error(error);
            }
            throw error;
        }
    }
    /**
     * @return {?}
     */
    save() {
        this.storage.setItem(this.storageKey, JSON.stringify(this.toObject()));
    }
    /**
     * @return {?}
     */
    restore() {
        if (!this.storage.getItem(this.storageKey)) {
            this.resetStorage(false);
            return;
        }
        try {
            const /** @type {?} */ sc = JSON.parse(this.storage.getItem(this.storageKey));
            if (!(sc.hasOwnProperty('items') && Array.isArray(sc.items) && sc.hasOwnProperty('taxRate') && sc.hasOwnProperty('shipping'))) {
                this.resetStorage('The object found under the key ' + this.storageKey + ' is not a valid cart object');
                return;
            }
            this._items = sc.items.map(i => {
                if (this.itemClass.fromJSON) {
                    return this.itemClass.fromJSON(i);
                }
                return new this.itemClass(i);
            });
            this.setTaxRate(parseFloat(sc.taxRate));
            this.setShipping(parseFloat(sc.shipping));
        }
        catch (/** @type {?} */ e) {
            this.resetStorage(e);
        }
    }
    /**
     * @param {?} item
     * @return {?}
     */
    addItem(item) {
        super.addItem(item);
        this.save();
    }
    /**
     * @param {?} id
     * @return {?}
     */
    removeItem(id) {
        super.removeItem(id);
        this.save();
    }
    /**
     * @return {?}
     */
    clear() {
        super.clear();
        this.save();
    }
    /**
     * @param {?} shipping
     * @return {?}
     */
    setShipping(shipping) {
        super.setShipping(shipping);
        this.save();
    }
    /**
     * @param {?} tax
     * @return {?}
     */
    setTaxRate(tax) {
        super.setTaxRate(tax);
        this.save();
    }
}
/** @nocollapse */
BrowserStorageCartService.ctorParameters = () => [
    { type: CartItem, decorators: [{ type: Inject, args: [CART_ITEM_CLASS,] },] },
    { type: undefined, decorators: [{ type: Inject, args: [CART_SERVICE_CONFIGURATION,] },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// unsupported: template constraints.
/**
 * An implementation of the cart service using localStorage to store items
 * \@order 5
 * @template T
 */
class LocalStorageCartService extends BrowserStorageCartService {
    /**
     * @param {?} itemClass
     * @param {?} configuration
     */
    constructor(itemClass, configuration) {
        super(itemClass, configuration) /* istanbul ignore next */;
        this.storage = window.localStorage;
        this.restore();
    }
}
LocalStorageCartService.ɵfac = function LocalStorageCartService_Factory(t) { return new (t || LocalStorageCartService)(ɵngcc0.ɵɵinject(CART_ITEM_CLASS), ɵngcc0.ɵɵinject(CART_SERVICE_CONFIGURATION)); };
LocalStorageCartService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: LocalStorageCartService, factory: LocalStorageCartService.ɵfac });
/** @nocollapse */
LocalStorageCartService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CART_ITEM_CLASS,] },] },
    { type: undefined, decorators: [{ type: Inject, args: [CART_SERVICE_CONFIGURATION,] },] },
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LocalStorageCartService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [CART_ITEM_CLASS]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [CART_SERVICE_CONFIGURATION]
            }] }]; }, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// unsupported: template constraints.
/**
 * An implementation of the cart service using sessionStorage to store items
 * \@order 4
 * @template T
 */
class SessionStorageCartService extends BrowserStorageCartService {
    /**
     * @param {?} itemClass
     * @param {?} configuration
     */
    constructor(itemClass, configuration) {
        super(itemClass, configuration) /* istanbul ignore next */;
        this.storage = window.sessionStorage;
        this.restore();
    }
}
SessionStorageCartService.ɵfac = function SessionStorageCartService_Factory(t) { return new (t || SessionStorageCartService)(ɵngcc0.ɵɵinject(CART_ITEM_CLASS), ɵngcc0.ɵɵinject(CART_SERVICE_CONFIGURATION)); };
SessionStorageCartService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: SessionStorageCartService, factory: SessionStorageCartService.ɵfac });
/** @nocollapse */
SessionStorageCartService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CART_ITEM_CLASS,] },] },
    { type: undefined, decorators: [{ type: Inject, args: [CART_SERVICE_CONFIGURATION,] },] },
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SessionStorageCartService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [CART_ITEM_CLASS]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [CART_SERVICE_CONFIGURATION]
            }] }]; }, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * A default implementation for CartItem
 *
 * \@order 2
 *
 * \@howToUse "Using properties and methods"
 * ```typescript
 * const item = new BaseCartItem({id: 1, name: 'Demo'});
 * item.quantity = 10;
 * item.setQuantity(50);
 * console.log(item.quantity) // prints 50
 * ```
 *
 * \@note {info} You can access item information either with direct property access or method calls, eg. `item.id === item.getId()`
 */
class BaseCartItem extends CartItem {
    /**
     * @param {?=} itemData
     */
    constructor(itemData = {}) {
        super();
        this.id = itemData.id || 0;
        this.name = itemData.name || '';
        this.price = itemData.price || 0;
        this.image = itemData.image || '';
        this.quantity = itemData.quantity || 1;
        this.data = itemData.data || {};
    }
    /**
     * Abstract base method implementation to obtain the item id
     * @return {?}
     */
    getId() {
        return this.id;
    }
    /**
     * Sets the current id for the item
     * @param {?} id {any}: The id value
     * @return {?}
     */
    setId(id) {
        this.id = id;
    }
    /**
     * Abstract base method implementation to return the name, a small text describing the item
     * @return {?}
     */
    getName() {
        return this.name;
    }
    /**
     * Sets the name of the item
     * @param {?} name
     * @return {?}
     */
    setName(name) {
        this.name = name;
    }
    /**
     * Abstract base method implementation to know how much the item cost
     * @return {?}
     */
    getPrice() {
        return this.price;
    }
    /**
     * Set the price of the item
     * @param {?} price
     * @return {?}
     */
    setPrice(price) {
        this.price = price;
    }
    /**
     * Abstract base method implementation to return how much of the item is ordered
     * @return {?}
     */
    getQuantity() {
        return this.quantity;
    }
    /**
     * Abstract base method implementation to set how much of the item is ordered
     * @param {?} quantity
     * @return {?}
     */
    setQuantity(quantity) {
        this.quantity = quantity;
    }
    /**
     * Abstract base method implementation to get the url of an image for the item
     * @return {?}
     */
    getImage() {
        return this.image;
    }
    /**
     * Sets the url of the item's image
     * @param {?} image
     * @return {?}
     */
    setImage(image) {
        this.image = image;
    }
    /**
     * Gets any additional data added to the item
     * @return {?}
     */
    getData() {
        return this.data;
    }
    /**
     * Sets any additional data to the item
     * @param {?} data
     * @return {?}
     */
    setData(data) {
        this.data = data;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @template T
 * @param {?} serviceType
 * @param {?} itemClass
 * @param {?} configuration
 * @return {?}
 */
function serviceFactory(serviceType, itemClass, configuration) {
    switch (serviceType) {
        case 'localStorage':
            return new LocalStorageCartService(itemClass, configuration);
        case 'sessionStorage':
            return new SessionStorageCartService(itemClass, configuration);
        default:
            return new MemoryCartService();
    }
}
/**
 * @param {?} serviceType
 * @return {?}
 */
function setupService(serviceType) {
    return {
        provide: CART_SERVICE_TYPE,
        useValue: serviceType || 'localStorage'
    };
}
/**
 * @param {?} itemClass
 * @return {?}
 */
function setItemClass(itemClass) {
    return {
        provide: CART_ITEM_CLASS,
        useValue: itemClass || BaseCartItem
    };
}
/**
 * @param {?} serviceType
 * @param {?} serviceOptions
 * @return {?}
 */
function setServiceConfiguration(serviceType, serviceOptions) {
    return {
        provide: CART_SERVICE_CONFIGURATION,
        useValue: serviceType !== 'memory' ? (!serviceOptions ? {
            storageKey: 'NgShoppingCart',
            clearOnError: true
        } : serviceOptions) : null
    };
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * A pipe that wraps the `CurrencyPipe` to set currency value display using a string rather than several arguments for easy configuration.
 *
 * \@summary
 * It takes a string as a single parameter in the format `'currencyCode:symbol:digitsInfo:locale'`. You can also use the special
 * value `'auto'` which will set the default used by Angular in that specific configuration.
 *
 * Every configuration is optional and not using any arguments or an argument of `'auto'`is equivalent to how the `CurrencyPipe` works by
 * default. If no locale is specified uses the current locale to format numbers.
 *
 * \@note {info} A value of `'auto:auto:auto:auto'` is equivalent to simply using `'auto'`.
 *
 * \@note {danger} In Angular versions lower than 6 the `CurrencyPipe` does not change the currency symbol if you don't specify a different
 * `currencyCode`
 *
 * \@howToUse "With a different currency symbol"
 * ```html
 * <span>
 *   {{ value | cartCurrency:format }}
 * </span>
 * ```
 * ```typescript
 * export class MyComponent {
 *   value = 10;
 *   format = 'EUR';
 * }
 * ```
 *
 * \@howToUse "With a five digits after the decimal point"
 * ```html
 * <span>
 *   {{ value | cartCurrency:format }}
 * </span>
 * ```
 * ```typescript
 * export class MyComponent {
 *   value = 10.56;
 *   format = 'auto:auto:1.5-5';
 * }
 * ```
 *
 * \@howToUse "With a different locale"
 * ```html
 * <span>
 *   {{ value | cartCurrency:format }}
 * </span>
 * ```
 * ```typescript
 * export class MyComponent {
 *   value = 10;
 *   format = 'auto:auto:auto:en-GB';
 * }
 * ```
 */
class CartCurrencyPipe {
    /**
     * @param {?} _locale
     */
    constructor(_locale) {
        this._locale = _locale;
        this.currencyFormatter = new CurrencyPipe(this._locale);
    }
    /**
     * @param {?} value
     * @param {?=} format
     * @return {?}
     */
    transform(value, format = 'auto') {
        if (!value && value !== 0) {
            return null;
        }
        const { currencyCode, display, digitsInfo, locale } = parseLocaleFormat(format);
        return this.currencyFormatter.transform(value, currencyCode, display, digitsInfo, locale);
    }
}
CartCurrencyPipe.ɵfac = function CartCurrencyPipe_Factory(t) { return new (t || CartCurrencyPipe)(ɵngcc0.ɵɵdirectiveInject(LOCALE_ID)); };
CartCurrencyPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "cartCurrency", type: CartCurrencyPipe, pure: true });
/** @nocollapse */
CartCurrencyPipe.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [LOCALE_ID,] },] },
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CartCurrencyPipe, [{
        type: Pipe,
        args: [{ name: 'cartCurrency' }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [LOCALE_ID]
            }] }]; }, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * The main exported library module. It includes `forRoot` and `forChild` static methods to support angular feature modules and singleton
 * services.
 *
 * \@note {danger} Only the `forRoot` method will configure providers for you. If you use the module without it you must configure the
 * library providers yourself.
 */
class ShoppingCartModule {
    /**
     * @param {?=} options
     * @return {?}
     */
    static forRoot(options = {}) {
        return {
            ngModule: ShoppingCartModule,
            providers: [
                setItemClass(options.itemType),
                setupService(options.serviceType),
                setServiceConfiguration(options.serviceType, options.serviceOptions),
                {
                    provide: CartService,
                    useFactory: serviceFactory,
                    deps: [CART_SERVICE_TYPE, CART_ITEM_CLASS, CART_SERVICE_CONFIGURATION]
                }
            ],
        };
    }
    /**
     * @return {?}
     */
    static forChild() {
        return {
            ngModule: ShoppingCartModule
        };
    }
}
ShoppingCartModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ShoppingCartModule });
ShoppingCartModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ShoppingCartModule_Factory(t) { return new (t || ShoppingCartModule)(); }, imports: [[
            CommonModule,
            HttpClientModule,
        ],
        CommonModule,
        HttpClientModule] });
/** @nocollapse */
ShoppingCartModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ShoppingCartModule, { declarations: function () { return [AddToCartEditorComponent,
        AddToCartComponent,
        CartCheckoutComponent,
        CartSummaryComponent,
        CartShowcaseComponent,
        CartViewComponent,
        ShowcaseOutletDirective,
        CartShowcaseItemComponent,
        CartCurrencyPipe]; }, imports: function () { return [CommonModule,
        HttpClientModule]; }, exports: function () { return [AddToCartEditorComponent,
        AddToCartComponent,
        CartCheckoutComponent,
        CartSummaryComponent,
        CartShowcaseComponent,
        CartViewComponent,
        CartShowcaseItemComponent,
        CartCurrencyPipe,
        CommonModule,
        HttpClientModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ShoppingCartModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    AddToCartEditorComponent,
                    AddToCartComponent,
                    CartCheckoutComponent,
                    CartSummaryComponent,
                    CartShowcaseComponent,
                    CartViewComponent,
                    ShowcaseOutletDirective,
                    CartShowcaseItemComponent,
                    CartCurrencyPipe,
                ],
                imports: [
                    CommonModule,
                    HttpClientModule,
                ],
                exports: [
                    AddToCartEditorComponent,
                    AddToCartComponent,
                    CartCheckoutComponent,
                    CartSummaryComponent,
                    CartShowcaseComponent,
                    CartViewComponent,
                    CartShowcaseItemComponent,
                    CartCurrencyPipe,
                    CommonModule,
                    HttpClientModule
                ],
                entryComponents: [CartShowcaseItemComponent]
            }]
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { AddToCartComponent, CartCheckoutComponent, CartSummaryComponent, CartShowcaseComponent, CartShowcaseItemComponent, CartViewComponent, CartCurrencyPipe, ShoppingCartModule, CART_ITEM_CLASS, CART_SERVICE_CONFIGURATION, CART_SERVICE_TYPE, CartService, MemoryCartService, LocalStorageCartService, SessionStorageCartService, CartItem, BaseCartItem, parseLocaleFormat, AddToCartEditorComponent as ɵa, ShowcaseOutletDirective as ɵb, serviceFactory as ɵc, setItemClass as ɵe, setServiceConfiguration as ɵf, setupService as ɵd, BrowserStorageCartService as ɵg };
